doozerlib/cli/release_gen_payload.py

```diff
@@ -59,762 +146,869 @@ def modify_and_replace_api_object(api_obj: oc.APIObject, modifier_func: Callable
-                release_tags: List = obj_model.spec['tags']
-                while len(release_tags) > 5:
-                    release_tags.pop(0)
-
-                # Now append a tag for our new nightly.
-                release_tags.append({
-                    'from': {
-                        'kind': 'DockerImage',
-                        'name': final_multi_pullspec,
-                    },
-                    'referencePolicy': {
-                        'type': 'Source'
-                    },
-                    'name': multi_release_name,
-                    'annotations': {
-                        'release.openshift.io/rewrite': 'false',  # Prevents the release controller from trying to create a local registry release payload with oc adm release new.
-                        # 'release.openshift.io/name': f'{runtime.get_minor_version()}.0-0.nightly',
-                    }
-                })
-                return True
-
-            modify_and_replace_api_object(multi_art_latest_is, add_multi_nightly_release, output_path, moist_run)
-
-    if not overall_permitted:
-        red_print('DO NOT PROCEED WITH THIS ASSEMBLY PAYLOAD -- not all detected issues are permitted.', file=sys.stderr)
-        exit(1)
+        return exchange_pullspec_tag_for_shasum(
+            multi_release_dest, find_manifest_list_sha(multi_release_dest))
+
+    def apply_multi_imagestream_update(self, final_multi_pullspec: str, imagestream_name: str, multi_release_name: str):
+        """Update the multi-release imagestream, pruning old nightlies and adding the new one."""
+        # [lmeyer] Q: but what if this is NOT a nightly? then we should not prune. but it's probably
+        # ok because we won't have more than one tag in non-nightly imagestreams?
+        multi_art_latest_is = self.ensure_imagestream_apiobj(imagestream_name)
+
+        def add_multi_nightly_release(obj: oc.APIObject):
+            obj_model = obj.model
+            if obj_model.spec.tags is oc.Missing:
+                obj_model.spec["tags"] = []
+
+            # For normal 4.x-art-latest, we update the imagestream with individual component images
+            # and the release controller formulates the nightly. For multi-arch, this is not
+            # possible (notably, the CI internal registry does not support manifest lists). Instead,
+            # in the ocp-multi namespace, the 4.x-art-latest imagestreams are configured
+            # `as: Stable`: https://github.com/openshift/release/pull/24130
+            # This means the release controller treats entries in these imagestreams the same way it
+            # treats it when ART tags into is/release; i.e. it treats it as an official release.
+            # With this comes the responsibility to prune nightlies ourselves.
+            release_tags: List = obj_model.spec["tags"]
+            while len(release_tags) > 5:  # keep at most 5 tags
```

Comment:
Per https://coreos.slack.com/archives/GDBRP5YJH/p1661329456656149, we only keep 1 tag for non-stream assemblies.